buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -545,11 +560,74 @@ static int writeUtf8(AbstractByteBuf buffer, int writerIndex, CharSequence seq,
         return writerIndex - oldWriterIndex;
     }
 
+    /**
+     * Returns max bytes length of UTF8 character sequence of the given length.
+     */
+    public static int utf8MaxBytes(final int seqLength) {
+        return seqLength * MAX_BYTES_PER_CHAR_UTF8;
+    }
+
     /**
      * Returns max bytes length of UTF8 character sequence.
+     * <p>
+     * It behaves like {@link #utf8MaxBytes(int)} applied to {@code seq} {@link CharSequence#length()}.
      */
     public static int utf8MaxBytes(CharSequence seq) {
-        return seq.length() * MAX_BYTES_PER_CHAR_UTF8;
+        return utf8MaxBytes(seq.length());
+    }
+
+    /**
+     * Returns the exact bytes length of UTF8 character sequence.
+     * <p>
+     * This method is producing the exact length according to {@link #writeUtf8(ByteBuf, CharSequence)}.
+     */
+    public static int utf8Bytes(final CharSequence seq) {
+        int seqLength = seq.length();
+        int i = 0;
+        // ASCII fast path
+        while (i < seqLength && seq.charAt(i) < 0x80) {
```

Comment:
optional: make 0x80 a char to avoid promotion.

Reply:
I'm not sure if you mean a promotion on Java or in the compiled ASM, so I've chosen the final source of truth, ASM :P
That's the `seq.charAt(i) < 0x80` part inside the unrolled loop:
```
0x00007f2ca0e5353b: nopl 0x0(%rax,%rax,1)  ;*aload_0
                                           ; - CompileTest::check@16 (line 26)
             L0001: movzwl 0x10(%r11,%rax,2),%r10d  ;*caload
                                                    ; - java.lang.String::charAt@27 (line 660)
                                                    ; - CompileTest::check@18 (line 26)
0x00007f2ca0e53546: cmp $0x80,%r10d
0x00007f2ca0e5354d: jge L0008  ;*if_icmpge
                               ; - CompileTest::check@26 (line 26)
0x00007f2ca0e5354f: movslq %eax,%r8
0x00007f2ca0e53552: movzwl 0x12(%r11,%r8,2),%r10d  ;*caload
                                                   ; - java.lang.String::charAt@27 (line 660)
                                                   ; - CompileTest::check@18 (line 26)
0x00007f2ca0e53558: cmp $0x80,%r10d
0x00007f2ca0e5355f: jge L0006  ;*if_icmpge
                               ; - CompileTest::check@26 (line 26)
0x00007f2ca0e53561: movzwl 0x14(%r11,%r8,2),%r10d  ;*caload
                                                   ; - java.lang.String::charAt@27 (line 660)
                                                   ; - CompileTest::check@18 (line 26)
0x00007f2ca0e53567: cmp $0x80,%r10d
0x00007f2ca0e5356e: jge L0007  ;*if_icmpge
                               ; - CompileTest::check@26 (line 26)
0x00007f2ca0e53570: movzwl 0x16(%r11,%r8,2),%r10d  ;*caload
                                                   ; - java.lang.String::charAt@27 (line 660)
                                                   ; - CompileTest::check@18 (line 26)
0x00007f2ca0e53576: cmp $0x80,%r10d
```
The only promotion I could spot here is `movzwl` that is zero extending to long the char/int returned by `seq.charAt`, but I can ensure you (because the ASM is the same) that putting a char cast or declaring `0x80` a char doesn't make difference: the ASM is 1:1 like this one.